### PR TITLE
Add davidson aglorithm to find out the lowest eigen values & vectors.

### PR DESCRIPTION
**Pull request by sheilaliuxl**

### CODE REVIEW

**Comment by sheilaliuxl** [comment]
@babbush @jarrodmcc 
Looks like the coverage is running infinitely, have you ever seen this/ do you have any idea why? Thanks!

File: src/openfermion/utils/_davidson.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+# Exceptions.
+class DavidsonError(Exception):
+    """Exceptions."""
+    pass
+
+
+class Davidson(object):
+    """Davidson algorithm to get the n states with smallest energies."""
+
+    def __init__(self, linear_operator, linear_operator_diagonal, eps=1e-6):
```
Comment (jarrodmcc) [inline src/openfermion/utils/_davidson.py]: If I recall correctly, Davidson also has another parameter which is the number of vectors to use in the subspace solves, should we pass that in here as well?

File: src/openfermion/utils/_davidson.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+    def get_lowest_n(self, n_lowest=1, initial_guess=None, max_iterations=300):
+        """
+        Returns `n` smallest eigenvalues and corresponding eigenvectors for
+            linear_operator.
+
+        Args:
+            n(int): The number of states corresponding to the smallest eigenvalues
+                and associated eigenvectors for the linear_operator.
+            initial_guess(numpy.ndarray[complex]): Initial guess of eigenvectors
```
Comment (jarrodmcc) [inline src/openfermion/utils/_davidson.py]: If I'm guessing right, this should have n columns, each of which is a guess at an eigenvector.  Do we need to put any constraints on this?  For example, do they need to be orthogonal for success?

We should maybe also check if initial_guess.shape[1] < n but not none, and pass a warning that we need to generate random vectors for the rest.

**Comment by babbush** [comment]
Looks like restarting the Travis build solved the coveralls problem.

File: src/openfermion/utils/_davidson.py (hunk 2 above)
Comment (sheilaliuxl) [reply src/openfermion/utils/_davidson.py]: 1. The only constraints is that all guess_v vectors are not zero. Added check.
2. Added random vectors in this case, though now really required.

File: src/openfermion/utils/_davidson.py (hunk 1 above)
Comment (sheilaliuxl) [reply src/openfermion/utils/_davidson.py]: Done, there was a TODO comment in the previous version, now removed.

**Comment by babbush** [comment]
Hi Sheila, usually, yes, we try to pass the coverage check.

**Comment by sheilaliuxl** [comment]
Now only one line is not covered, which is kind of expected, since it should be very (or even extremely) rare to see all new directions are linearly dependent with original vectors.

**Pull request merged**